skyportal/facility_apis/lt.py

```diff
@@ -0,0 +1,325 @@
+from datetime import datetime
+
+from astropy.coordinates import SkyCoord
+from astropy import units as u
+from astropy.io import ascii
+
+from . import FollowUpAPI
+from baselayer.app.access import auth_or_token
+from baselayer.app.env import load_env
+import requests
+from requests_toolbelt.utils import dump
+
+
+LT_SETTINGS = {
+    'USERNAME': '',
+    'PASSWORD': '',
+    'LT_HOST': '',
+    'LT_PORT': '',
+    'MAX_AIRMASS': 2.0,
+    'MAX_SEEING': 1.2,
+    'MAX_SKYBRI': 1.0,
+    'PHOTOMETRIC': 'light',
+    'IOO_BINNING': '2x2',
+}
+
+env, cfg = load_env()
+
+LT_SETTINGS['USERNAME'] = cfg['app.lt_username']
+LT_SETTINGS['PASSWORD'] = cfg['app.lt_password']
+LT_SETTINGS['LT_HOST'] = cfg['app.lt_host']
+LT_SETTINGS['LT_PORT'] = cfg['app.lt_port']
+
+LT_XML_NS = 'http://www.rtml.org/v3.1a'
+LT_XSI_NS = 'http://www.w3.org/2001/XMLSchema-instance'
+LT_SCHEMA_LOCATION = 'http://www.rtml.org/v3.1a http://telescope.livjm.ac.uk/rtml/RTML-nightly.xsd'
+
+
+class LTRequest():
+
+    def is_valid(self):
+        super().is_valid()
+        errors = LTFacility.validate_observation(self, self.observation_payload())
+        if errors:
+            self.add_error(None, errors)
+        return not errors
+
+    def _build_prolog(self):
+        namespaces = {
+            'xsi': LT_XSI_NS,
+        }
+        schemaLocation = etree.QName(LT_XSI_NS, 'schemaLocation')
+        uid = format(str(int(time.time())))
+        prolog = etree.Element('RTML', {schemaLocation: LT_SCHEMA_LOCATION},
+                               xmlns=LT_XML_NS,
+                               mode='request', uid=uid,
+                               version='3.1a', nsmap=namespaces)
+        return prolog
+
+    def _build_project(self, payload, request):
+        project = etree.Element('Project',
+                                ProjectID=request.payload["LT_proposalID"])
+        contact = etree.SubElement(project, 'Contact')
+        etree.SubElement(contact, 'Username').text = LT_SETTINGS['USERNAME']
+        etree.SubElement(contact, 'Name').text = LT_SETTINGS['USERNAME']
+        payload.append(project)
+
+    def _build_constraints(self, request):
+        airmass_const = etree.Element('AirmassConstraint', maximum=str(LT_SETTINGS['MAX_AIRMASS']))
+
+        sky_const = etree.Element('SkyConstraint')
+        etree.SubElement(sky_const, 'Flux').text = str(LT_SETTINGS['MAX_SKYBRI'])
+        etree.SubElement(sky_const, 'Units').text = 'magnitudes/square-arcsecond'
+
+        seeing_const = etree.Element('SeeingConstraint',
+                                     maximum=(str(LT_SETTINGS['MAX_SEEING'])),
+                                     units='arcseconds')
+
+        photom_const = etree.Element('ExtinctionConstraint')
+        etree.SubElement(photom_const, 'Clouds').text = LT_SETTINGS['PHOTOMETRIC']
+
+        date_const = etree.Element('DateTimeConstraint', type='include')
+        start = request.payload["start_date"] + 'T00:00:00+00:00'
```

Comment:
I think the start/end dates are already ISO timestamps?

Reply:
@dannygoldstein I don't think so... <class 'str'>...